### PR TITLE
Start auto-reloading prow secrets.

### LINKED ISSUE #4761: Autoreload Prow secrets.
While debugging, I autogenerated a new oauth token for my Prow cluster. This invalidated the previous token (which was working). I added the new token to the Prow cluster secrets, but hook jobs were getting Bad Credentials errors when trying to run presubmit on PRs.

Bouncing the Hook and Plank jobs fixed the issue.

### PR DESCRIPTION
**Pull request by droslean**
Fixes #4761 
Implements secretAgent.
Follow-up issues are needed for making the rest components to use secretAgent and for fixing the graphql library for the tide.

### CODE REVIEW

**Comment by droslean** [comment]
wip because some unit tests are missing.

File: prow/config/secrets_agent.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+					}
+					lastModTime = recentModTime
+				}
+
+				if sc, err := LoadSecrets(paths...); err != nil {
+					logrus.WithField("Secret paths", paths).
+						WithError(err).Error("Error loading config.")
+				} else {
+					skips = 0
+					sa.Lock()
```
Comment (stevekuznetsov) [inline prow/config/secrets_agent.go]: Could use `Set()` here

File: prow/config/secrets_agent.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+func (sa *SecretAgent) Start(paths ...string) error {
+	sc, err := LoadSecrets(paths...)
+	if err != nil {
+		return err
+	}
+	sa.sc = sc
+
+	go func() {
```
Comment (stevekuznetsov) [inline prow/config/secrets_agent.go]: Why not use `fsnotify` here with a timer backup? Did we ever move to that for the config agent @BenTheElder

File: prow/cmd/hook/main.go [hunk 3]
```diff
@@ -109,17 +110,25 @@ func main() {
...
 
-	webhookSecretRaw, err := ioutil.ReadFile(o.webhookSecretFile)
+	secretAgent := &config.SecretAgent{}
+	if err := secretAgent.Start(o.webhookSecretFile, o.githubTokenFile); err != nil {
+		logrus.WithError(err).Fatal("Error starting secrets agent.")
+	}
+
+	sc := secretAgent.Config()
+
+	webhookSecret := bytes.TrimSpace(sc.SecretsMap[o.webhookSecretFile])
```
Comment (stevekuznetsov) [inline prow/cmd/hook/main.go]: Seems like we could trim space before storing so that end users of the agent don't need to do this

File: prow/cmd/hook/main.go [hunk 4]
```diff
@@ -110,34 +108,30 @@ func main() {
 		logrus.WithError(err).Fatal("Error starting config agent.")
 	}
 
+	var tokens []string
+	var err error
```
Comment (stevekuznetsov) [inline prow/cmd/hook/main.go]: I *think* this is unused? In the following:

```go
if err := somefunc(); err != nil {
```

The `err` does not escape the `if` block scope

File: prow/cmd/hook/main.go [hunk 5]
```diff
@@ -110,34 +108,30 @@ func main() {
...
+
+	// Append the path of hmac and github secrets.
+	tokens = append(tokens, o.githubTokenFile)
```
Comment (stevekuznetsov) [inline prow/cmd/hook/main.go]:
```go
tokens := []string{
    o.githubTokenFile,
    o.webhookSecretFile,
}
```

File: prow/cmd/hook/main.go [hunk 6]
```diff
@@ -110,34 +108,30 @@ func main() {
...
+	tokens = append(tokens, o.webhookSecretFile)
+
+	// This is necessary because there is no default value for slack token file.
```
Comment (stevekuznetsov) [inline prow/cmd/hook/main.go]: technically necessary since the slack client is optional (which is why there is no default, unlike for the other tokens, which are mandatory)

File: prow/cmd/hook/main.go [hunk 7]
```diff
@@ -110,34 +108,30 @@ func main() {
...
-
-	var teamToken string
-	if o.slackTokenFile != "" {
-		teamTokenRaw, err := ioutil.ReadFile(o.slackTokenFile)
-		if err != nil {
-			logrus.WithError(err).Fatal("Could not read slack token file.")
-		}
-		teamToken = string(bytes.TrimSpace(teamTokenRaw))
-	}
+	fmt.Printf("Slack token %s \n", teamToken)
```
Comment (stevekuznetsov) [inline prow/cmd/hook/main.go]: Should be removed?

File: prow/cmd/hook/main.go [hunk 8]
```diff
@@ -109,27 +107,29 @@ func main() {
 		logrus.WithError(err).Fatal("Error starting config agent.")
 	}
 
-	webhookSecretRaw, err := ioutil.ReadFile(o.webhookSecretFile)
-	if err != nil {
-		logrus.WithError(err).Fatal("Could not read webhook secret file.")
-	}
-	webhookSecret := bytes.TrimSpace(webhookSecretRaw)
+	var tokens []string
+	var err error
```
Comment (stevekuznetsov) [inline prow/cmd/hook/main.go]: Where is this `err` used?

File: prow/cmd/hook/main.go [hunk 9]
```diff
@@ -109,27 +107,29 @@ func main() {
...
-	if err != nil {
-		logrus.WithError(err).Fatal("Could not read oauth secret file.")
-	}
-	oauthSecret := string(bytes.TrimSpace(oauthSecretRaw))
+	// Append the path of hmac and github secrets.
+	tokens = append(tokens, o.githubTokenFile)
+	tokens = append(tokens, o.webhookSecretFile)
 
-	var teamToken string
+	// This is necessary because there is no default value for slack token file.
```
Comment (stevekuznetsov) [inline prow/cmd/hook/main.go]: This is necessary since slack token is optional

File: prow/config/config.go [hunk 10]
```diff
@@ -36,12 +37,23 @@ import (
...
-	"k8s.io/test-infra/prow/github"
 	"k8s.io/test-infra/prow/kube"
 	"k8s.io/test-infra/prow/pod-utils/decorate"
 	"k8s.io/test-infra/prow/pod-utils/downwardapi"
 )
 
+// PullRequestMergeType enumerates the types of merges the GitHub API can
+// perform
+// https://developer.github.com/v3/pulls/#merge-a-pull-request-merge-button
+type PullRequestMergeType string
```
Comment (stevekuznetsov) [inline prow/config/config.go]: How is this related?

File: prow/config/config.go [hunk 11]
```diff
@@ -36,12 +37,23 @@ import (
...
+
+// Possible types of merges for the GitHub merge API
+const (
+	MergeMerge  PullRequestMergeType = "merge"
```
Comment (stevekuznetsov) [inline prow/config/config.go]: For autocomplete we try to use `PullRequestMergeTypeMerge` etc

File: prow/config/config.go [hunk 12]
```diff
@@ -362,6 +374,30 @@ func loadConfig(prowConfig, jobConfig string) (*Config, error) {
 	return nc, nil
 }
 
+// Loads multiple paths of secrets and add them in a map.
+func LoadSecrets(paths []string) (map[string][]byte, error) {
+	secretsMap := make(map[string][]byte)
```
Comment (stevekuznetsov) [inline prow/config/config.go]: If you're going to use `make` here you can size it as well to the length of the paths

File: prow/config/config.go [hunk 13]
```diff
@@ -362,6 +374,30 @@ func loadConfig(prowConfig, jobConfig string) (*Config, error) {
...
+
+	for _, path := range paths {
+		stat, err := os.Stat(path)
+		if err != nil {
+			return nil, err
+		}
+
+		if stat.IsDir() {
```
Comment (stevekuznetsov) [inline prow/config/config.go]: Why do we need a specific callout for this and the `os.Stat` call? Won't `ioutil.ReadFIle` fail in those cases and we can just bubble up that error?

File: prow/config/secrets_agent.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+// the failure message but continue attempting to load.
+func (sa *SecretAgent) Start(paths []string) error {
+	secretsMap, err := LoadSecrets(paths)
+	if err != nil {
+		return err
+	}
+
+	sa.secretsMap = secretsMap
+
+	go func() {
```
Comment (stevekuznetsov) [inline prow/config/secrets_agent.go]: Why not `fsnotify`? Have we added that watch to the config agent, @BenTheElder

File: prow/config/tide.go [hunk 15]
```diff
@@ -80,7 +79,7 @@ type Tide struct {
 
 	// A key/value pair of an org/repo as the key and merge method to override
 	// the default method of merge. Valid options are squash, rebase, and merge.
-	MergeType map[string]github.PullRequestMergeType `json:"merge_method,omitempty"`
+	MergeType map[string]PullRequestMergeType `json:"merge_method,omitempty"`
```
Comment (stevekuznetsov) [inline prow/config/tide.go]: Did you have a cycle? This seems unrelated to your changes

File: prow/github/client.go [hunk 16]
```diff
@@ -387,7 +394,16 @@ func (c *Client) doRequest(method, path, accept string, body interface{}) (*http
 	if err != nil {
 		return nil, err
 	}
-	req.Header.Set("Authorization", "Token "+c.token)
+
+	var token string
+	if c.secretAgent != nil {
```
Comment (stevekuznetsov) [inline prow/github/client.go]: When is it valid for the agent to be `nil`? If it's for testing, let's not allow that and use a fake agent

File: prow/github/client.go [hunk 17]
```diff
@@ -187,15 +190,17 @@ func (c *Client) Throttle(hourlyTokens, burst int) {
 //   An endpoint is used when all preceding endpoints have returned a conn err.
 //   This should be used when using the ghproxy GitHub proxy cache to allow
 //   this client to bypass the cache if it is temporarily unavailable.
-func NewClient(token string, bases ...string) *Client {
+func NewClient(token string, secretPath string, secretAgent *config.SecretAgent, bases ...string) *Client {
```
Comment (stevekuznetsov) [inline prow/github/client.go]: This API is a bit unclear -- you give the token as read once, then the agent and the path. I don't think it's necessary for the GitHub client to need to know about the inner workings of the secret agent, so maybe we can instead pass `func() (string)` that could generate the secret at run-time as necessary:

```go
func whatever() {
    agent := NewSecretAgent(tokens)
    client := NewClient(func() string {
        return agent.Config()[githubToken]
    }, bases...)
}

func NewClient(token func() string, bases ...string) *Client {
    ...
}
```

File: prow/hook/server.go [hunk 18]
```diff
@@ -39,6 +39,8 @@ type Server struct {
 	Plugins     *plugins.PluginAgent
 	ConfigAgent *config.Agent
 	HMACSecret  []byte
```
Comment (stevekuznetsov) [inline prow/hook/server.go]: Should not store this in the struct

**Comment by droslean** [comment]
@stevekuznetsov Thanks for the review. I am new to prow, I will check all your comments and let you know

File: prow/config/tide.go (hunk 15 above)
Comment (droslean) [reply prow/config/tide.go]: Yes, there was a cycle, and those constants are used only in config package

File: prow/github/client.go (hunk 16 above)
Comment (droslean) [reply prow/github/client.go]: For start, only hook is using the secretAgent, so the rest are using the token as they used to. That's why I am giving nil secretAgent from the rest binaries. This should be removed when all the binaries will use the secretAgent.

File: prow/config/secrets_agent.go (hunk 14 above)
Comment (droslean) [reply prow/config/secrets_agent.go]: +1. fsnotify should be the best solution. In secretAgent should be easier to implement than the config agent. I will check that out.

**Comment by kargakis** [comment]
fsnotify is orthogonal to this change and should be handled in a different
PR if at all

**Comment by droslean** [comment]
Please don't mind the multiple commits. I will squash them, once the PR is ready to be merged.

File: prow/config/config.go (hunk 12 above)
Comment (droslean) [reply prow/config/config.go]: How much length should be ok? The hook is using 3 paths. Is there any other binary that is using more?

File: prow/config/config.go (hunk 10 above)
Comment (stevekuznetsov) [reply prow/config/config.go]: We can't move this type into `config` from `kube`. Must find a different way to thread things through.

File: prow/config/config_test.go [hunk 19]
```diff
@@ -761,3 +761,58 @@ periodics:
 		}
 	}
 }
+
+// Integration test for fake secrets loading in a secret agent.
+// Checking also if the agent changes the secret's values as expected.
+func TestSecreAgentLoading(t *testing.T) {
```
Comment (stevekuznetsov) [inline prow/config/config_test.go]: `s/TestSecre/TestSecret/`

File: prow/config/config_test.go [hunk 20]
```diff
@@ -761,3 +761,58 @@ periodics:
...
+	secretsMap := secretAgent.Config()
+
+	// Check if the values are as expected.
+	for k, v := range secretsMap {
+		if string(v) != tempTokenValue {
+			t.Fatalf("In secret %s it was expected %s but found %s", k, tempTokenValue, string(v))
+		}
+	}
+
+	// Change the values
```
Comment (stevekuznetsov) [inline prow/config/config_test.go]: Don't you want to change the files?

File: prow/github/client.go (hunk 17 above)
Comment (stevekuznetsov) [reply prow/github/client.go]: This is the largest point of feedback

File: prow/github/client.go (hunk 16 above)
Comment (stevekuznetsov) [reply prow/github/client.go]: Let's get them all migrated in one pull, please

File: prow/config/secrets_agent.go [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+					sa.Unlock()
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+// Config returns the latest config. Do not modify the config.
+func (sa *SecretAgent) Config() map[string][]byte {
```
Comment (stevekuznetsov) [inline prow/config/secrets_agent.go]: Not sure if this is the best API here -- getting a map back implies you can do something with it, even if the comment says no. Maybe we should have

```go
func (sa *SecretAgent) Get(secret string) []byte {}
```

To make it clear it's read-only?

File: prow/config/secrets_agent.go [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+	sa.Lock()
+	defer sa.Unlock()
+	return sa.secretsMap
+}
+
+// Set sets the config. Useful for testing.
+func (sa *SecretAgent) Set(secretsMap map[string][]byte) {
```
Comment (stevekuznetsov) [inline prow/config/secrets_agent.go]: If this is only for testing, can we use an interface for the SecretAgent and have this method only exit on the test mock? There is no case for production code to set the secret data, right?

File: prow/config/config_test.go (hunk 19 above)
Comment (droslean) [reply prow/config/config_test.go]: oops :)

File: prow/config/config_test.go (hunk 20 above)
Comment (droslean) [reply prow/config/config_test.go]: Yes actually, I want to test how the agent is interacting to changes of the files.

File: prow/config/secrets_agent.go (hunk 22 above)
Comment (droslean) [reply prow/config/secrets_agent.go]: After the secret will change, the agent will set the new data. I didn't use Set() there, but I changed it after your suggestion.

File: prow/config/secrets_agent.go (hunk 21 above)
Comment (droslean) [reply prow/config/secrets_agent.go]: You are right, I refactored this. The name GetSecrets is more accurate. Also comments updated because it's not read-only.

File: label_sync/main.go [hunk 23]
```diff
@@ -548,16 +548,16 @@ func newClient(tokenPath string, tokens, tokenBurst int, dryRun bool, hosts ...s
...
+
+	secretAgent := config.NewSecretAgent()
+	if err := secretAgent.Start([]string{tokenPath}); err != nil {
+		logrus.WithError(err).Fatal("Error starting secrets agent.")
 	}
-	oauthSecret := string(bytes.TrimSpace(b))
 
 	if dryRun {
-		return github.NewDryRunClient(oauthSecret, hosts...), nil
+		return github.NewDryRunClient(tokenPath, secretAgent, hosts...), nil
```
Comment (stevekuznetsov) [inline label_sync/main.go]: Please refactor to use the generator

File: experiment/manual-trigger/manual-trigger.go [hunk 24]
```diff
@@ -100,39 +102,45 @@ func sanityCheckFlags(o options) error {
...
 	ac := jenkins.AuthConfig{}
 	if o.jenkinsTokenFile != "" {
-		token, err := loadToken(o.jenkinsTokenFile)
-		if err != nil {
-			log.Fatalf("cannot read file specified by --jenkins-token-file: %v", err)
-		}
 		ac.Basic = &jenkins.BasicAuthConfig{
 			User:  o.jenkinsUserName,
-			Token: token,
+			Token: getSecret(secretAgent, o.jenkinsTokenFile),
```
Comment (stevekuznetsov) [inline experiment/manual-trigger/manual-trigger.go]: Doesn't this mean that we load the Jenkins token just once? We should defer loading to when we need it or use a goroutine to keep a local copy in sync if the former approach is too resource intensive on every jenkins api call

File: prow/.gitignore [hunk 25]
```diff
@@ -13,3 +13,9 @@ cmd/jenkins-operator/jenkins-operator
 cmd/tide/tide
 cmd/clonerefs/clonerefs
 cmd/initupload/initupload
+cmd/phony
```
Comment (stevekuznetsov) [inline prow/.gitignore]: This makes me think you're not using the `bazel` workflow -- which is fine, but at some point you might want to use it, the rest of the repo is very `bazel`-heavy

File: prow/config/config_test.go [hunk 26]
```diff
@@ -921,5 +921,59 @@ func TestBrancher_Intersects(t *testing.T) {
...
+
+	// Check if the values are as expected.
+	for _, tempSecret := range tempSecrets {
+		tempSecretValue := secretAgent.GetSecret(tempSecret)
+		if string(tempSecretValue) != tempTokenValue {
+			t.Fatalf("In secret %s it was expected %s but found %s", tempSecret, tempTokenValue, tempSecretValue)
+		}
+	}
+
+	// Change the values
```
Comment (stevekuznetsov) [inline prow/config/config_test.go]: Would still expect you to actually change the files instead of calling `Set` here

File: prow/config/tide.go (hunk 15 above)
Comment (stevekuznetsov) [reply prow/config/tide.go]: This is outstanding, we should not move github types out of the package. Should determine a different way to structure this to avoid the cycle

Comment (stevekuznetsov) [reply prow/config/tide.go]: A separate package declaring an interface that the config impl can satisfy is often the approach

File: prow/git/git.go [hunk 27]
```diff
@@ -95,17 +98,21 @@ func (c *Client) SetRemote(remote string) {
 
 // SetCredentials sets credentials in the client to be used for pushing to
 // or pulling from remote repositories.
-func (c *Client) SetCredentials(user, pass string) {
+func (c *Client) SetCredentials(user string, tokenGenerator func() []byte) {
 	c.credLock.Lock()
 	defer c.credLock.Unlock()
 	c.user = user
-	c.pass = pass
```
Comment (stevekuznetsov) [inline prow/git/git.go]: should remove the pass field then

File: prow/git/git.go [hunk 28]
```diff
@@ -95,17 +98,21 @@ func (c *Client) SetRemote(remote string) {
...
+	c.tokenGenerator = tokenGenerator
 }
 
 func (c *Client) getCredentials() (string, string) {
 	c.credLock.RLock()
 	defer c.credLock.RUnlock()
-	return c.user, c.pass
+	if c.tokenGenerator != nil {
```
Comment (stevekuznetsov) [inline prow/git/git.go]: this should be impossible with the constructor

File: prow/github/client.go [hunk 29]
```diff
@@ -191,14 +194,14 @@ func (c *Client) Throttle(hourlyTokens, burst int) {
 //   An endpoint is used when all preceding endpoints have returned a conn err.
 //   This should be used when using the ghproxy GitHub proxy cache to allow
 //   this client to bypass the cache if it is temporarily unavailable.
-func NewClient(token string, bases ...string) *Client {
+func NewClient(token func() []byte, bases ...string) *Client {
 	return &Client{
 		logger: logrus.WithField("client", "github"),
 		time:   &standardTime{},
-		gqlc:   githubql.NewClient(oauth2.NewClient(context.Background(), oauth2.StaticTokenSource(&oauth2.Token{AccessToken: token}))),
+		gqlc:   githubql.NewClient(oauth2.NewClient(context.Background(), oauth2.StaticTokenSource(&oauth2.Token{AccessToken: string(token())}))),
```
Comment (stevekuznetsov) [inline prow/github/client.go]: This reads the token only once

File: prow/github/client.go [hunk 30]
```diff
@@ -391,7 +394,13 @@ func (c *Client) doRequest(method, path, accept string, body interface{}) (*http
 	if err != nil {
 		return nil, err
 	}
-	req.Header.Set("Authorization", "Token "+c.token)
+
+	if c.token != nil {
```
Comment (stevekuznetsov) [inline prow/github/client.go]: nil is not allowed

File: prow/config/tide.go (hunk 15 above)
Comment (droslean) [reply prow/config/tide.go]: I will check that approach, and move them back to github package.

File: prow/github/client.go (hunk 29 above)
Comment (droslean) [reply prow/github/client.go]: I will try to find the solution to pass the agent to graphql

File: prow/github/client.go (hunk 30 above)
Comment (droslean) [reply prow/github/client.go]: I added this check to satisfy the tests.

File: experiment/manual-trigger/manual-trigger.go [hunk 31]
```diff
@@ -20,15 +20,17 @@ package main
...
 	"net/url"
 	"strings"
 
+	"k8s.io/test-infra/prow/config"
 	"k8s.io/test-infra/prow/github"
 	"k8s.io/test-infra/prow/jenkins"
 	"k8s.io/test-infra/prow/kube"
 	"k8s.io/test-infra/prow/pod-utils/downwardapi"
+
+	"github.com/sirupsen/logrus"
```
Comment (stevekuznetsov) [inline experiment/manual-trigger/manual-trigger.go]: nit: this import should be in it's own group after stdlib but before `k8s.io/test-infra`

File: experiment/manual-trigger/manual-trigger.go (hunk 24 above)
Comment (stevekuznetsov) [reply experiment/manual-trigger/manual-trigger.go]: This is still outstanding?

File: prow/cmd/jenkins-operator/main.go [hunk 32]
```diff
@@ -230,14 +240,6 @@ func main() {
...
-func loadToken(file string) (string, error) {
-	raw, err := ioutil.ReadFile(file)
-	if err != nil {
-		return "", err
-	}
-	return string(bytes.TrimSpace(raw)), nil
-}
-
 func loadCerts(certFile, keyFile, caCertFile string) (*tls.Config, error) {
 	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
```
Comment (kargakis) [inline prow/cmd/jenkins-operator/main.go]: Open a follow-up issue to add support for x509 certs in the secret agent.

File: prow/cmd/plank/main.go [hunk 33]
```diff
@@ -101,37 +99,35 @@ func main() {
...
 	var pkcs map[string]*kube.Client
 	if o.dryRun {
-		if oauthSecret != "" {
-			ghc = github.NewDryRunClient(oauthSecret, o.githubEndpoint.Strings()...)
-		}
+		ghc = github.NewDryRunClient(getSecret(secretAgent, o.githubTokenFile), o.githubEndpoint.Strings()...)
 		kc = kube.NewFakeClient(o.deckURL)
 		pkcs = map[string]*kube.Client{kube.DefaultClusterAlias: kc}
 	} else {
-		if oauthSecret != "" {
```
Comment (kargakis) [inline prow/cmd/plank/main.go]: The github client still needs to be set conditionally here.

File: prow/config/secrets_agent.go (hunk 14 above)
Comment (kargakis) [reply prow/config/secrets_agent.go]: follow-up issue even though I don't feel strong about fsnotify

File: prow/git/git.go [hunk 34]
```diff
@@ -45,6 +45,9 @@ type Client struct {
 	// pass is used when pushing or pulling code if specified.
 	pass string
```
Comment (kargakis) [inline prow/git/git.go]: Needs to be removed?

File: prow/git/git.go (hunk 28 above)
Comment (kargakis) [reply prow/git/git.go]: Seems outstanding

File: prow/github/client.go (hunk 29 above)
Comment (kargakis) [reply prow/github/client.go]: Follow-up issue for this. Most likely the solution is to update the SecretAgent to implement [TokenSource](https://godoc.org/golang.org/x/oauth2#TokenSource).

File: prow/github/client.go (hunk 30 above)
Comment (kargakis) [reply prow/github/client.go]: Add the token in the tests that require it vs allowing the separate path here.

File: prow/jenkins/jenkins.go [hunk 35]
```diff
@@ -208,13 +208,13 @@ type AuthConfig struct {
...
-	User  string
-	Token string
+	User           string
+	TokenGenerator func() []byte
 }
 
 // BearerTokenAuthConfig authenticates jenkins using an oauth bearer token.
 type BearerTokenAuthConfig struct {
-	Token string
+	TokenGenerator func() []byte
```
Comment (kargakis) [inline prow/jenkins/jenkins.go]:
```go
GetToken func() []byte
```
since the generation piece is opaque for the client.

File: prow/github/client.go [hunk 36]
```diff
@@ -63,11 +63,11 @@ type Client struct {
 
 	gqlc     gqlClient
 	client   httpClient
-	token    string
 	bases    []string
 	dry      bool
 	fake     bool
 	throttle throttler
+	token    func() []byte
```
Comment (kargakis) [inline prow/github/client.go]: `s/token/getToken/`

File: prow/cmd/jenkins-operator/main.go (hunk 32 above)
Comment (droslean) [reply prow/cmd/jenkins-operator/main.go]: opened #8701

File: prow/github/client.go (hunk 29 above)
Comment (droslean) [reply prow/github/client.go]: opened #8702

File: prow/cmd/branchprotector/protect.go [hunk 37]
```diff
@@ -108,17 +107,23 @@ func main() {
 		logrus.WithError(err).Fatalf("Failed to load --config=%s", o.config)
```
Comment (kargakis) [inline prow/cmd/branchprotector/protect.go]: Can you also fix --config to --config-path since you also fixed the above?

**Comment by kargakis** [comment]
lgtm, I think you can squash now

**Comment by kargakis** [comment]
/lgtm

@cjwagner can you approve?

**Comment by stevekuznetsov** [comment]
IIRC Cole is on PTO but we will wait for his return to merge.

**Comment by stevekuznetsov** [comment]
/lgtm
@cjwagner

File: experiment/manual-trigger/manual-trigger.go [hunk 38]
```diff
@@ -100,39 +102,45 @@ func sanityCheckFlags(o options) error {
...
+	if o.jenkinsBearerTokenFile != "" {
+		tokens = append(tokens, o.jenkinsBearerTokenFile)
+	}
+
+	secretAgent := config.NewSecretAgent()
+	if err := secretAgent.Start(tokens); err != nil {
+		logrus.WithError(err).Fatal("Error starting secrets agent.")
+	}
+
+	getSecret := func(secretAgent *config.SecretAgent, secretPath string) func() []byte {
```
Comment (cjwagner) [inline experiment/manual-trigger/manual-trigger.go]: nit: None of the `getSecret` functions require the `secretAgent *config.SecretAgent` parameter. `secretAgent` is already captured by the anonymous function closure.

File: prow/.gitignore (hunk 25 above)
Comment (cjwagner) [inline prow/.gitignore]: cmd/phony/phony

File: prow/.gitignore [hunk 39]
```diff
@@ -13,3 +13,9 @@ cmd/jenkins-operator/jenkins-operator
...
+cmd/peribolos
```
Comment (cjwagner) [inline prow/.gitignore]: cmd/peribolos/peribolos

File: prow/config/secrets_agent.go [hunk 40]
```diff
@@ -0,0 +1,101 @@
...
+
+// SecretAgent watches a path and automatically loads the secrets stored.
+type SecretAgent struct {
+	sync.Mutex
+	secretsMap map[string][]byte
+}
+
+// NewSecretAgent returns a SecretAgent with a map of strings with []byte value.
+// The key will be used to index the file path, and the value to store the secret value.
+func NewSecretAgent() *SecretAgent {
```
Comment (cjwagner) [inline prow/config/secrets_agent.go]: nit: This could be combined with `Start`. I don't think there is any case in which we would construct a `SecretAgent` that we don't intend to immediately use.

File: prow/config/secrets_agent.go [hunk 41]
```diff
@@ -0,0 +1,101 @@
...
+					}
+					lastModTime = recentModTime
+				}
+
+				if secretsMap, err := LoadSecrets(paths); err != nil {
+					logrus.WithField("Secret paths", paths).
+						WithError(err).Error("Error loading config.")
+				} else {
+					skips = 0
+					sa.Set(secretsMap)
```
Comment (cjwagner) [inline prow/config/secrets_agent.go]: We should break here. We just loaded the secrets so we don't need to check the other values in `secretsMap`.

File: prow/config/secrets_agent.go [hunk 42]
```diff
@@ -0,0 +1,101 @@
...
+					if !recentModTime.After(lastModTime) {
+						skips++
+						continue // file hasn't been modified
+					}
+					lastModTime = recentModTime
+				}
+
+				if secretsMap, err := LoadSecrets(paths); err != nil {
+					logrus.WithField("Secret paths", paths).
+						WithError(err).Error("Error loading config.")
```
Comment (cjwagner) [inline prow/config/secrets_agent.go]: Trying to load all of the secrets when we find that one secret has updated doesn't seem like the right approach, especially if we continue on to check the other secrets immediately after reloading them all.
Additionally, with this pattern, if one secret fails to update we will fail to update all other secrets as well.

File: prow/github/client_test.go [hunk 43]
```diff
@@ -48,8 +48,14 @@ func (tt *testTime) Until(t time.Time) time.Duration {
 }
 
 func getClient(url string) *Client {
+	getToken := func() func() []byte {
```
Comment (cjwagner) [inline prow/github/client_test.go]: There is an unneeded level of indirection here (an extra functional layer).

File: prow/github/client.go [hunk 44]
```diff
@@ -186,40 +186,42 @@ func (c *Client) Throttle(hourlyTokens, burst int) {
...
 	return &Client{
 		logger: logrus.WithField("client", "github"),
 		time:   &standardTime{},
-		gqlc:   githubql.NewClient(oauth2.NewClient(context.Background(), oauth2.StaticTokenSource(&oauth2.Token{AccessToken: token}))),
-		client: &http.Client{},
-		token:  token,
-		bases:  bases,
-		dry:    false,
+		gqlc: githubql.NewClient(oauth2.NewClient(context.Background(),
+			oauth2.StaticTokenSource(&oauth2.Token{AccessToken: string(getToken())}))),
```
Comment (cjwagner) [inline prow/github/client.go]: This calls `getToken` only once during client creation, but we need it to be called every time a githubql client call is made in order to pick up any changes to the token.

File: prow/prstatus/prstatus.go [hunk 45]
```diff
@@ -222,8 +222,15 @@ func (da *DashboardAgent) HandlePrStatus(queryHandler PullRequestQueryHandler) h
 				serverError("Save oauth session", err)
 				return
 			}
+
+			getSecret := func() func() []byte {
```
Comment (cjwagner) [inline prow/prstatus/prstatus.go]: Extra functional layer here.

File: prow/github/client.go (hunk 44 above)
Comment (droslean) [reply prow/github/client.go]: another issue is already opened for it. #8702

File: prow/cmd/hook/main.go [hunk 46]
```diff
@@ -109,41 +107,45 @@ func main() {
 		logrus.WithError(err).Fatal("Error starting config agent.")
 	}
 
-	webhookSecretRaw, err := ioutil.ReadFile(o.webhookSecretFile)
-	if err != nil {
-		logrus.WithError(err).Fatal("Could not read webhook secret file.")
+	var err error
+	for _, ep := range o.githubEndpoint.Strings() {
+		_, err = url.Parse(ep)
```
Comment (cjwagner) [inline prow/cmd/hook/main.go]: ParseRequestURI?

File: prow/config/config.go [hunk 47]
```diff
@@ -371,6 +373,30 @@ func loadConfig(prowConfig, jobConfig string) (*Config, error) {
...
 
+// LoadSecrets loads multiple paths of secrets and add them in a map.
+func LoadSecrets(paths []string) (map[string][]byte, error) {
+	secretsMap := make(map[string][]byte, len(paths))
+
+	for _, path := range paths {
+		b, err := ioutil.ReadFile(path)
+		if err != nil {
+			return nil, fmt.Errorf("error reading %s: %v", path, err)
+		}
```
Comment (cjwagner) [inline prow/config/config.go]: Use `LoadSingleSecret` instead of copy-pasting.

File: prow/config/secrets_agent.go [hunk 48]
```diff
@@ -0,0 +1,96 @@
...
+			recentModTime := secretStat.ModTime()
+			if !recentModTime.After(lastModTime) {
+				skips++
+				continue // file hasn't been modified
+			}
+			lastModTime = recentModTime
+		}
+
+		if secretValue, err := LoadSingleSecret(secretPath); err != nil {
+			logrus.WithField("Secret path: ", secretPath).
```
Comment (cjwagner) [inline prow/config/secrets_agent.go]: Just create the `logrus.Entry` once at the top of the function and use it throughout instead of reconstructing for every use.
As is you are using field keys `Secret path: ` and `secretPath` which will not be easily searchable because they don't match. Please use the key `secret-path` for consistency with the rest of our log field keys.

File: prow/cmd/hook/main.go [hunk 49]
```diff
@@ -109,41 +107,45 @@ func main() {
...
-	for _, ep := range o.githubEndpoint.Strings() {
-		_, err = url.Parse(ep)
-		if err != nil {
-			logrus.WithError(err).Fatalf("Invalid --endpoint URL %q.", ep)
+	secretAgent := &config.SecretAgent{}
+	if err := secretAgent.Start(tokens); err != nil {
+		logrus.WithError(err).Fatal("Error starting secrets agent.")
+	}
+
+	teamToken := string(secretAgent.GetSecret(o.slackTokenFile))
```
Comment (stevekuznetsov) [inline prow/cmd/hook/main.go]: Why is this read just once?

File: prow/cmd/plank/main.go [hunk 50]
```diff
@@ -100,37 +98,47 @@ func main() {
...
-			if err != nil {
-				logrus.WithError(err).Fatalf("Invalid --endpoint URL %q.", ep)
+	var ghc plank.GitHubClient
+	if o.githubTokenFile != "" {
+		secretAgent := &config.SecretAgent{}
+		if err := secretAgent.Start([]string{o.githubTokenFile}); err != nil {
+			logrus.WithError(err).Fatal("Error starting secrets agent.")
+		}
+
+		getSecret := func(secretPath string) func() []byte {
```
Comment (stevekuznetsov) [inline prow/cmd/plank/main.go]: Almost feels like we should expose a secret getter from the agent since you've written this a lot

File: prow/config/config_test.go [hunk 51]
```diff
@@ -921,5 +921,83 @@ func TestBrancher_Intersects(t *testing.T) {
...
+		counter := 0
+		for counter <= retries {
+			tempSecretValue := secretAgent.GetSecret(tempSecret)
+			if string(tempSecretValue) != changedTokenValue {
+				if counter == retries {
+					errors = append(errors, fmt.Sprintf("In secret %s it was expected %s but found %s\n",
+						tempSecret, changedTokenValue, tempSecretValue))
+				} else {
+					// Secret agent needs some time to update the values. So wait and retry.
+					time.Sleep(100 * time.Millisecond)
```
Comment (stevekuznetsov) [inline prow/config/config_test.go]: This might be fragile

File: prow/config/secrets_agent.go [hunk 52]
```diff
@@ -0,0 +1,97 @@
...
+			recentModTime := secretStat.ModTime()
+			if !recentModTime.After(lastModTime) {
+				skips++
+				continue // file hasn't been modified
+			}
+			lastModTime = recentModTime
+		}
+
+		if secretValue, err := LoadSingleSecret(secretPath); err != nil {
+			logger.WithField("secret-path: ", secretPath).
```
Comment (stevekuznetsov) [inline prow/config/secrets_agent.go]: no colon necessary, these become JSON keys

File: prow/cmd/hook/main.go (hunk 49 above)
Comment (droslean) [reply prow/cmd/hook/main.go]: opened follow-up issue #8843

File: prow/config/config_test.go (hunk 51 above)
Comment (droslean) [reply prow/config/config_test.go]: Another approach is to dramatically increase the retries value and remove that.
Also, this is only used for tests.
